File: lexicon-android/app/src/main/java/com/robertvokac/lexicon/ui/graph/GraphLayout.kt
package com.robertvokac.lexicon.ui.graph

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

/**
 * Places the nodes of a relationship graph: rings by distance from the centre
 * to start with, then a Fruchterman-Reingold pass that pushes every pair apart
 * and pulls linked items together. Deterministic, and the same steps as the
 * desktop (lexicon-qt/GraphLayout.h) and web (js/graphlayout.js) clients.
 */
object GraphLayout {
    data class Point(val x: Double, val y: Double)

    private const val RING = 150.0
    private const val IDEAL = 120.0
    private const val ITERATIONS = 250

    /** How the graph sits on a canvas: pixels per layout unit, and the layout point at the canvas centre. */
    data class Fit(val scale: Double, val centreX: Double, val centreY: Double)

    /**
     * Fits every point into [width] x [height] pixels around the first (centre)
     * point, with [margin] pixels for circles and captions. The first point
     * stays in the screen centre even when the other nodes are asymmetrical.
     */
    fun fit(points: List<Point>, width: Int, height: Int, margin: Double, maxScale: Double): Fit {
        if (points.isEmpty() || width <= 0 || height <= 0) return Fit(maxScale, 0.0, 0.0)
        val centreX = points.first().x
        val centreY = points.first().y
        val extentX = points.maxOf { kotlin.math.abs(it.x - centreX) }
        val extentY = points.maxOf { kotlin.math.abs(it.y - centreY) }
        val roomX = max(1.0, width - 2 * margin)
        val roomY = max(1.0, height - 2 * margin)
        val scaleX = if (extentX > 0) roomX / (2 * extentX) else maxScale
        val scaleY = if (extentY > 0) roomY / (2 * extentY) else maxScale
        return Fit(min(min(scaleX, scaleY), maxScale), centreX, centreY)
    }

    /** [depths] of node i from the centre, node 0; [edges] as index pairs. */
    fun layout(depths: List<Int>, edges: List<Pair<Int, Int>>): List<Point> {
        val count = depths.size
        val xs = DoubleArray(count)
        val ys = DoubleArray(count)
        val perDepth = IntArray(4)
        val seen = IntArray(4)
        depths.forEach { perDepth[it.coerceIn(0, 3)]++ }
        for (i in 1 until count) {
            val depth = depths[i].coerceIn(1, 3)
            val angle = 2 * PI * seen[depth]++ / max(1, perDepth[depth]) + depth * 0.7
            xs[i] = RING * depth * cos(angle)
            ys[i] = RING * depth * sin(angle)
        }
        val shiftX = DoubleArray(count)
        val shiftY = DoubleArray(count)
        repeat(ITERATIONS) { step ->
            val temperature = 80.0 * (1.0 - step.toDouble() / ITERATIONS) + 2.0
            shiftX.fill(0.0)
            shiftY.fill(0.0)
            for (i in 0 until count) {
                for (j in i + 1 until count) {
                    val dx = xs[i] - xs[j]
                    val dy = ys[i] - ys[j]
                    val distance = max(hypot(dx, dy), 0.01)
                    val force = IDEAL * IDEAL / distance
                    shiftX[i] += dx / distance * force; shiftY[i] += dy / distance * force
                    shiftX[j] -= dx / distance * force; shiftY[j] -= dy / distance * force
                }
            }
            for ((a, b) in edges) {
                if (a == b) continue
                val dx = xs[a] - xs[b]
                val dy = ys[a] - ys[b]
                val distance = max(hypot(dx, dy), 0.01)
                val force = distance * distance / IDEAL
                shiftX[a] -= dx / distance * force; shiftY[a] -= dy / distance * force
                shiftX[b] += dx / distance * force; shiftY[b] += dy / distance * force
            }
            for (i in 1 until count) { // The centre stays where it is.
                val length = hypot(shiftX[i], shiftY[i])
                if (length < 1e-9) continue
                val move = min(length, temperature)
                xs[i] += shiftX[i] / length * move
                ys[i] += shiftY[i] / length * move
            }
        }
        return List(count) { Point(xs[it], ys[it]) }
    }
}

File: lexicon-web/js/graphlayout.js
// Places the nodes of a relationship graph: rings by distance from the centre
// to start with, then a Fruchterman-Reingold pass that pushes every pair apart
// and pulls linked items together. Deterministic, and the same steps as the
// desktop (lexicon-qt/GraphLayout.h) and Android clients. No DOM, so it also
// runs under Node for its tests.

// depths[i] is node i's distance from the centre, node 0; edges are [a, b]
// index pairs. Returns [{ x, y }] with the centre at the origin.
export function layoutGraph(depths, edges) {
    const count = depths.length;
    const RING = 150;
    const IDEAL = 120;
    const ITERATIONS = 250;
    const perDepth = [0, 0, 0, 0];
    const seen = [0, 0, 0, 0];
    for (const depth of depths) perDepth[Math.min(3, Math.max(0, depth))] += 1;
    const points = depths.map(() => ({ x: 0, y: 0 }));
    for (let i = 1; i < count; i += 1) {
        const depth = Math.min(3, Math.max(1, depths[i]));
        const angle = (2 * Math.PI * seen[depth]) / Math.max(1, perDepth[depth]) + depth * 0.7;
        seen[depth] += 1;
        points[i] = { x: RING * depth * Math.cos(angle), y: RING * depth * Math.sin(angle) };
    }
    for (let step = 0; step < ITERATIONS; step += 1) {
        const temperature = 80 * (1 - step / ITERATIONS) + 2;
        const shift = points.map(() => ({ x: 0, y: 0 }));
        for (let i = 0; i < count; i += 1) {
            for (let j = i + 1; j < count; j += 1) {
                const dx = points[i].x - points[j].x;
                const dy = points[i].y - points[j].y;
                const distance = Math.max(Math.hypot(dx, dy), 0.01);
                const force = (IDEAL * IDEAL) / distance;
                shift[i].x += (dx / distance) * force;
                shift[i].y += (dy / distance) * force;
                shift[j].x -= (dx / distance) * force;
                shift[j].y -= (dy / distance) * force;
            }
        }
        for (const [a, b] of edges) {
            if (a === b) continue;
            const dx = points[a].x - points[b].x;
            const dy = points[a].y - points[b].y;
            const distance = Math.max(Math.hypot(dx, dy), 0.01);
            const force = (distance * distance) / IDEAL;
            shift[a].x -= (dx / distance) * force;
            shift[a].y -= (dy / distance) * force;
            shift[b].x += (dx / distance) * force;
            shift[b].y += (dy / distance) * force;
        }
        for (let i = 1; i < count; i += 1) { // The centre stays where it is.
            const length = Math.hypot(shift[i].x, shift[i].y);
            if (length < 1e-9) continue;
            const move = Math.min(length, temperature);
            points[i].x += (shift[i].x / length) * move;
            points[i].y += (shift[i].y / length) * move;
        }
    }
    return points;
}
